altamisa/isatab/models_validation.py

```diff
@@ -0,0 +1,289 @@
+                        msg = tpl.format(
+                            test,
+                            self.protocols[process.protocol_ref].type.name,
+                            ", ".join(protocol_type_restrictions[test]),
+                        )
+                        msgs.append(msg)
+                if msgs:
+                    msg = "\n".join(msgs)
+            if msg:
+                warnings.warn(msg, IsaValidationWarning)
+
+    def _validate_name_types(self, process: models.Process):
+        # Match restricted name types to corresponding assay technologies and protocols
+        self._validate_restrictions(
+            process.name_type, process, RESTRICTED_TECH_NAMES, RESTRICTED_PROTO_NAMES
+        )
+
+    def _validate_special_case_annotations(self, process: models.Process):
+        # Match restricted annotations to corresponding assay technologies and protocols
+        if process.array_design_ref:
+            self._validate_restrictions(
+                headers.ARRAY_DESIGN_REF, process, RESTRICTED_TECH_ANNOS, RESTRICTED_PROTO_ANNOS
+            )
+        if process.first_dimension or process.second_dimension:
+            self._validate_restrictions(
+                headers.FIRST_DIMENSION, process, RESTRICTED_TECH_ANNOS, RESTRICTED_PROTO_ANNOS
+            )
+            self._validate_restrictions(
+                headers.SECOND_DIMENSION, process, RESTRICTED_TECH_ANNOS, RESTRICTED_PROTO_ANNOS
+            )
+
+
+class ArcValidator:
+    """Validator for Arcs"""
+
+    def __init__(self, assay: models.AssayInfo = None):
+        self.assay = (assay,)
+
+    def validate(self, arc: models.Arc):
+        """Run Arc validations"""
+        # Nothing to do yet
+        # TODO: Move all non-interruptive arc validations here
+        pass
+
+    # TODO: In studies, check that source only start arcs and that samples only end arcs
+    # TODO: In assays, check that samples only start arcs
+
+
+class _AssayAndStudyValidator:
+    """Base validator for Study and Assay"""
+
+    def validate(self):
+        self._validate_materials()
+        self._validate_processes()
+        self._validate_arcs()
+
+
+class StudyValidator(_AssayAndStudyValidator):
+    """Validator for Study"""
+
+    def __init__(self, study_info: models.StudyInfo, study: models.Study):
+        self.study_info = study_info
+        self.study = study
+
+    def _validate_materials(self):
+        # Iterate materials and validate
+        validator = MaterialValidator()
+        [validator.validate(m) for m in self.study.materials.values()]
+
+    def _validate_processes(self):
+        # Iterate processes and validate
+        validator = ProcessValidator(self.study_info.protocols)
+        [validator.validate(p) for p in self.study.processes.values()]
+
+    def _validate_arcs(self):
+        # Iterate arcs and validate
+        validator = ArcValidator()
+        [validator.validate(a) for a in self.study.arcs]
```

Comment:
?! List comprehensions should only be used when actually returning something, not for side effects.

Reply:
same above